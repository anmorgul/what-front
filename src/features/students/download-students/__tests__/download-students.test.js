import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter as Router, useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';

import { DownloadStudents } from '../download-students.js';
import { paths, useActions } from '@/shared';
import {
  importStudentsMockData,
  loadStudentGroupsMockData,
} from '../mocks/mock-data.js';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

jest.mock('@/shared/hooks', () => ({ useActions: jest.fn([jest.fn()]) }));

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useHistory: jest.fn().mockReturnValue({
    push: jest.fn(),
  }),
}));

const useActionsFns = {
  sendStudents: jest.fn(),
  globalLoadsStudentsGroups: jest.fn(),
};

useActions.mockReturnValue([
  useActionsFns.sendStudents,
  useActionsFns.globalLoadsStudentsGroups,
]);

describe('DownloadStudents', () => {
  const history = useHistory();
  let formValues;

  beforeEach(() => {
    useSelector
      .mockReturnValue(loadStudentGroupsMockData)
      .mockReturnValue(importStudentsMockData);

    formValues = {
      groupId: '1',
      students: '../mocks/mock-file.xlsx',
    };
  });

  it('should render component', () => {
    const { container } = render(
      <Router>
        <DownloadStudents />
      </Router>
    );
    const downloadStudentsContainer = container.getElementsByClassName(
      'container'
    );

    expect(downloadStudentsContainer).toMatchSnapshot();
  });

  it('should navigate groups page if groups are loaded', () => {
    render(
      <Router>
        <DownloadStudents />
      </Router>
    );

    expect(history.push).toHaveBeenCalledWith(paths.GROUPS);
  });

  it('should validate the form', async () => {
    const { getByText } = render(
      <Router>
        <DownloadStudents />
      </Router>
    );

    const submitButton = getByText('Send');

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    expect(submitButton).toBeDisabled();
  });

  it('should submit the form with correct schema', async () => {
    const { getByRole, getByLabelText, getByText } = render(
      <Router>
        <DownloadStudents />
      </Router>
    );

    const [downloadStudents] = useActions(useActionsFns.sendStudents);

    const groupId = getByRole('combobox');
    const students = getByLabelText('Choose File');
    const submitButton = getByText('Send');

    fireEvent.change(groupId, {
      target: {
        value: formValues.groupId,
      },
    });

    fireEvent.change(students, {
      target: {
        files: formValues.students,
      },
    });

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    downloadStudents({
      groupId: groupId.value,
      students: students.files,
    });

    expect(downloadStudents).toHaveBeenCalled();
  });
});
