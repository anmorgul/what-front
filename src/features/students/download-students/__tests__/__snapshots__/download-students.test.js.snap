// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadStudents should render component 1`] = `
HTMLCollection [
  <div
    class="container pt-5"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="w-100 card shadow p-4 ml-2 mr-2"
      >
        <div
          class="px-2 py-4"
        >
          <h3>
            Upload Student('s)
          </h3>
          <hr />
          <form
            action="#"
          >
            <div
              class="row m-0 pt-3"
            >
              <div
                class="col-md-4 font-weight-bolder"
              >
                <label
                  for="groupsInput"
                >
                  Group:
                </label>
              </div>
              <div
                class="col-md-8"
              >
                <select
                  class="custom-select"
                  id="groupsInput"
                  name="groupId"
                />
              </div>
            </div>
            <div
              class="row m-0 pt-3"
            >
              <div
                class="col-md-4 font-weight-bolder pt-3"
              >
                <label
                  for="file"
                >
                  Student('s):
                </label>
              </div>
              <div
                class="col-md-8"
              >
                <input
                  accept=".xlsx"
                  class="form-control"
                  hidden=""
                  id="actual-btn"
                  name="students"
                  type="file"
                />
                <label
                  class="mr-2 col-md-6"
                  for="actual-btn"
                >
                  Choose File
                </label>
                <span
                  class="font-weight-bolder"
                />
              </div>
            </div>
            <div
              class="row justify-content-end px-4"
            >
              <button
                class="btn btn-info w-25"
                disabled=""
                type="submit"
              >
                Send
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;
